tests/Microsoft.Identity.Test.Unit.net45/Integration/DeviceCodeFlowIntegrationTest.cs

```diff
@@ -0,0 +1,169 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Test.LabInfrastructure;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Chrome;
+using OpenQA.Selenium.Support.UI;
+
+namespace Microsoft.Identity.Test.Unit.Integration
+{
+    // Important: do not install a NuGet package with the Chrome driver as it is a security risk.
+    // Instead, install the Chrome driver on the test machine
+
+    // Note: these tests require permission to a KeyVault Microsoft account; 
+    // Please ignore them if you are not a Microsoft FTE, they will run as part of the CI build
+    [TestClass]
+    public class DeviceCodeFlow
+    {
+        private static readonly string[] Scopes = { "User.Read" };
+
+        public TestContext TestContext { get; set; }
+
+        [ClassInitialize]
+        public static void ClassInitialize(TestContext context)
+        {
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+        }
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            TestCommon.ResetState();
+        }
+
+
+        [TestMethod]
+        [Timeout(60 * 1000)] // 1 min timeout
+        public async Task DeviceCodeFlowTestAsync()
+        {
+            Debug.WriteLine("Fetching user from lab");
+            LabResponse labResponse = LabUserHelper.GetDefaultUser();
+
+            Debug.WriteLine("Calling AcquireTokenWithDeviceCodeAsync");
+            PublicClientApplication pca = new PublicClientApplication(labResponse.AppId);
+            var result = await pca.AcquireTokenWithDeviceCodeAsync(Scopes, deviceCodeResult =>
+            {
+                RunAutomatedDeviceCodeFlow(deviceCodeResult, labResponse.User);
+
+                return Task.FromResult(0);
+            }).ConfigureAwait(false);
+
+            Assert.IsNotNull(result);
+            Assert.IsTrue(!string.IsNullOrEmpty(result.AccessToken));
+        }
+
+        private void RunAutomatedDeviceCodeFlow(DeviceCodeResult deviceCodeResult, LabUser user)
+        {
+            IWebDriver driver = null;
+            try
+            {
+                driver = InitDriver();
+
+                Debug.WriteLine("Navigating and enterign the code");
+
+                driver.Navigate().GoToUrl(deviceCodeResult.VerificationUrl);
+                driver.FindElement(By.Id("code")).SendKeys(deviceCodeResult.UserCode);
+
+                IWebElement continueBtn = WaitForElementToBeEnabled(driver, By.Id("continueBtn"));
+                continueBtn?.Click();
+
+                Debug.WriteLine("Logging in");
```

Comment:
>Debug.WriteLine("Logging in"); [](start = 16, length = 30)

If we want this pattern, such lines should be in the helper method and not in each test case. not sure we need those but if it helps with debugging I suppose it's good.

Reply:
done